deepchem/models/torch_models/chemberta.py

```diff
@@ -0,0 +1,174 @@
+
+        else:
+            self.config = RobertaConfig(
+                vocab_size=vocab_size,
+                max_position_embeddings=max_position_embedddings,
+                num_attention_heads=number_attention_heads,
+                num_hidden_layers=num_hidden_layers,
+                type_vocab_size=type_vocab_size,
+            )
+
+            self.model = RobertaForMaskedLM(config=self.config)
+            print(f"Model size: {self.model.num_parameters()} parameters.")
+
+            if tokenizer_type == 0: # generate novel BPE tokenizer for dataset
+                from transformers import ByteLevelBPETokenizer
+
+                tokenizer_path = tokenizer_output_dir
+                if not os.path.isdir(tokenizer_path):
+                    os.makedirs(tokenizer_path)
+
+                tokenizer = ByteLevelBPETokenizer()
+                tokenizer.train(files=dataset_path, vocab_size=vocab_size, min_frequency=BPE_min_frequency, special_tokens=["<s>","<pad>","</s>","<unk>","<mask>"])
+                tokenizer.save_model(tokenizer_path)
+            else: # just assign huggingface hub model path
+                tokenizer_path = model_path
+                print ('load ST tokenizer')
+
+        self.tokenizer = RobertaTokenizerFast.from_pretrained(tokenizer_path, max_tokenizer_len)
+
+class ChemBERTaMaskedLMModel(TorchModel):
+    def __init__(self,
+                vocab_size: int = 600,
+                max_position_embeddings = 515,
+                number_attention_heads=12,
+                num_hidden_layers=6,
+                dataset_path: str = '',
+                mode='pre-trained',
+                model_path='DeepChem/SmilesTokenizer_PubChem_1M',
+                mlm_probability: float = 0.15,
+                output_dir = 'model/',
+                frac_train: int = 0.95,
+                eval_steps: int = 1000,
+                logging_steps: int = 100,
+                overwrite_output_dir: boolean = True,
+                num_train_epochs: int = 10,
+                per_device_train_batch_size: int = 64,
+                save_steps :int = 10000,
+                save_total_limit: int = 2,
+                fp16: bool= True,
+                run_name: str = 'chemberta_lm'
+                **kwargs):
+        
+        from transformers import LineByLineTextDataset
+        from transformers import DataCollatorForLanguageModeling
+        from transformers import TrainingArguments
+
+        model = ChemBERTa(
+            vocab_size=vocab_size,
+            max_position_embeddings = max_position_embeddings,
+            number_attention_heads = number_attention_heads,
+            num_hidden_layers = num_hidden_layers,
+            dataset_path = dataset_path,
+            mode = mode,
+            model_path = model_path,
+            model_output_dir = model_output_dir,
+            **kwargs)
+                
+        dataset = LineByLineTextDataset(file_path=dataset_path, 
```

Comment:
We should see if we can use a DeepChem dataset instead. Also, the dataset shouldn't be constructed in the constructor

Reply:
Sounds good - I'll try playing around with diskdataset and see if I can get it working! I'll also move over the dataset code into _fit()

Reply:
How important is the reliance on the `Trainer` class? If parts of the code can be used from `Trainer` (like DataLoader creation moved to `default_generator`, and we have a similar loop in `fit_generator`, then that would simplify things.

Reply:
The [trainer class](https://huggingface.co/transformers/_modules/transformers/trainer.html#Trainer) might be a considerable amount of technical debt to inherit + maintain - mainly because all of the Huggingface model infra revolve around passing into a trainer object. We use the trainer class to pass shards from the dataset into the transformer, enable mixed precision (fp16) training as done in the chemberta paper, compute masked loss during training and push to the Deepchem model hub.  I can try to use pure PyTorch code instead, but I think it may be a lot less stable and easier to maintain. 

What do you guys think? If I can try to get an end-to-end Chemberta deepchem integration going (i.e - feature done, models for pre-training + fine-tuning, DiskDataset integration, etc), we can go back and improve the compatibility if that works!

Reply:
I think using the trainer class under the hood seems fine so long as we can have the external API of the class match DeepChem's usual API. Is the code in this PR ready for a second round of review?